### PR TITLE
Feature/namespace permissions

### PR DESCRIPTION
**Pull request by pfeuffer**
## Proposed changes

Adds permission settings known from individual repositories to namespace level.

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] PR is well described
- [x] Related issues linked to PR if existing and labels set
- [x] Target branch is not master (in most cases develop should bet the target of choice) 
- [x] Code does not conflict with target branch
- [x] New code is covered with unit tests
- [x] CHANGELOG.md updated
- [x] Definition of Done's fulfilled: [DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/definition-of-done.md) // [UI DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/ui-dod.md)
- [x] Documentation updated (only necessary for new features or changed behaviour)

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

**Review by eheimbuch — CHANGES_REQUESTED** [review]
some unit tests are failing

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+      .filter(filterPermission(permissionName))
+      .findFirst()
+      .orElseThrow(() -> notFound(entity(RepositoryPermission.class, permissionName).in(Namespace.class, namespaceName)));
+    RepositoryPermission newPermission = dtoToModelMapper.map(permission);
+    if (!namespace.removePermission(existingPermission)) {
+      throw new IllegalStateException(String.format("could not delete modified permission %s from namespace %s", existingPermission, namespaceName));
+    }
+    namespace.addPermission(newPermission);
+    manager.modify(namespace);
+    log.info("the permission with name: {} is updated.", permissionName);
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java:259]: Do we really want to log before and after each time? Seems a bit much logging.

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+      .ifPresent(permission -> {
+        namespace.removePermission(permission);
+        manager.modify(namespace);
+      });
+    log.info("the permission with name: {} is deleted.", permissionName);
+  }
+
+  private Predicate<RepositoryPermission> filterPermission(String name) {
+    return permission -> getPermissionName(name).equals(permission.getName())
+      &&
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java]: ugly line break

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/RepositoryPermissionToRepositoryPermissionDtoMapper.java [hunk 3]
```diff
@@ -51,18 +53,19 @@
...
-   * Add the self, update and delete links.
-   *
-   * @param target     the mapped dto
-   * @param repository the repository
-   */
+  @BeforeMapping
+  void validatePermissions(@Context Namespace namespace) {
+    NamespacePermissions.permissionRead().check();
+  }
+
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/api/v2/resources/RepositoryPermissionToRepositoryPermissionDtoMapper.java:68]: Should the mapper do to the permission check? I thought it should happen earlier. Even the resource seems a better place for this.

File: scm-webapp/src/main/java/sonia/scm/repository/DefaultNamespaceManager.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      .orElseThrow(() -> notFound(entity(Namespace.class, namespace.getNamespace())));
+    fireEvent(HandlerEventType.BEFORE_MODIFY, namespace, oldNamespace);
+    dao.add(namespace);
+    fireEvent(HandlerEventType.MODIFY, namespace, oldNamespace);
+  }
+
+  @Subscribe
+  public void cleanupDeletedNamespaces(RepositoryEvent repositoryEvent) {
+    HandlerEventType eventType = repositoryEvent.getEventType();
+    if (eventType == HandlerEventType.DELETE || eventType == HandlerEventType.MODIFY && !repositoryEvent.getItem().getNamespace().equals(repositoryEvent.getOldItem().getNamespace())) {
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/repository/DefaultNamespaceManager.java:84]: move this long check into own method(s)

File: scm-webapp/src/main/resources/locales/en/plugins.json [hunk 5]
```diff
@@ -93,6 +93,16 @@
         "description": "May see and manage all installed and available plugins"
       }
     },
+    "namespace": {
+      "permissionRead": {
+        "displayName": "read permissions on namespaces",
```
Comment (eheimbuch) [inline scm-webapp/src/main/resources/locales/en/plugins.json:98]: Start Displayname with Capital letter.

File: scm-webapp/src/main/resources/locales/en/plugins.json [hunk 6]
```diff
@@ -93,6 +93,16 @@
...
+        "description": "May see the permissions set for namespaces"
+      },
+      "permissionWrite": {
+        "displayName": "modify permissions on namespaces",
```
Comment (eheimbuch) [inline scm-webapp/src/main/resources/locales/en/plugins.json:102]: Start Displayname with Capital letter.

**Review by eheimbuch — CHANGES_REQUESTED** [review]

File: scm-ui/ui-types/src/Repositories.ts [hunk 7]
```diff
@@ -58,5 +58,6 @@ export type NamespaceCollection = {
 
 export type RepositoryGroup = {
   name: string;
+  namespace: Namespace;
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-types/src/Repositories.ts:61]: Well, yeah, but does anybody "implement" this?

File: scm-ui/ui-webapp/public/locales/de/repos.json [hunk 8]
```diff
@@ -218,7 +218,8 @@
   "renameRepo": {
     "button": "Repository umbenennen",
     "subtitle": "Benennt dieses Repository um",
-    "description": "Es werden keine Weiterleitung auf den neuen Namen eingerichtet.",
+    "description1": "Es werden keine Weiterleitung auf den neuen Namen eingerichtet.",
+    "description2": "Berechtigungen aus dem Namespace werden bei einem Wechsel nicht übernommen.",
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/public/locales/de/repos.json:238]: Think that I've missed a case for auth cache invalidation here. Gotta do this when the namespace changes.

File: scm-ui/ui-webapp/src/repos/components/list/RepositoryGroupEntry.tsx [hunk 9]
```diff
@@ -32,10 +34,23 @@ type Props = {
 class RepositoryGroupEntry extends React.Component<Props> {
   render() {
     const { group } = this.props;
+    const settingsLink = group.namespace?._links?.permissions && (
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/src/repos/components/list/RepositoryGroupEntry.tsx:37]: Would be helpful, yes

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java (hunk 2 above)
Comment (pfeuffer) [reply scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java]: Stumbled upon this already by myself, you're right

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/RepositoryPermissionToRepositoryPermissionDtoMapper.java (hunk 3 above)
Comment (pfeuffer) [reply scm-webapp/src/main/java/sonia/scm/api/v2/resources/RepositoryPermissionToRepositoryPermissionDtoMapper.java:68]: Right. This one was a "copy" of the existing logic, though it is not necessary here. Will remove this.

File: scm-webapp/src/main/java/sonia/scm/repository/DefaultNamespaceManager.java (hunk 4 above)
Comment (pfeuffer) [reply scm-webapp/src/main/java/sonia/scm/repository/DefaultNamespaceManager.java:84]: :+1:

File: scm-ui/ui-types/src/Repositories.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-types/src/Repositories.ts:61]: Okay, got this :-) Will change it.

File: scm-ui/ui-webapp/src/repos/components/list/groupByNamespace.ts [hunk 10]
```diff
@@ -58,3 +63,7 @@ function sortByName(a, b) {
   }
   return 0;
 }
+
+function findNamespace(namespaces: NamespaceCollection, namespaceToFind: string) {
+  return namespaces._embedded.namespaces.filter(namespace => namespace.namespace === namespaceToFind)[0];
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/src/repos/components/list/groupByNamespace.ts:68]: Thx for the hint ‒ was looking for this somehow...

File: scm-ui/ui-webapp/src/repos/namespaces/containers/PermissionsNavLink.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+import { WithTranslation, withTranslation } from "react-i18next";
+import { Namespace } from "@scm-manager/ui-types";
+import { NavLink } from "@scm-manager/ui-components";
+
+type Props = WithTranslation & {
+  permissionUrl: string;
+  namespace: Namespace;
+};
+
+class PermissionsNavLink extends React.Component<Props> {
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/src/repos/namespaces/containers/PermissionsNavLink.tsx:34]: Next time...

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+import static sonia.scm.NotFoundException.notFound;
+import static sonia.scm.api.v2.resources.RepositoryPermissionDto.GROUP_PREFIX;
+
+@Slf4j
+public class NamespacePermissionResource {
+
+  private RepositoryPermissionDtoToRepositoryPermissionMapper dtoToModelMapper;
+  private RepositoryPermissionToRepositoryPermissionDtoMapper modelToDtoMapper;
+  private RepositoryPermissionCollectionToDtoMapper repositoryPermissionCollectionToDtoMapper;
+  private ResourceLinks resourceLinks;
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java:66]: Wonder why I've missed these... thx

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java (hunk 1 above)
Comment (pfeuffer) [reply scm-webapp/src/main/java/sonia/scm/api/v2/resources/NamespacePermissionResource.java:259]: No, definitely not, you're right. Removed the first one.

File: scm-core/src/main/java/sonia/scm/repository/Namespace.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+  public Namespace(String namespace) {
+    this.namespace = namespace;
+  }
+
+  /**
+   * Constructor for JaxB, only.
+   */
+  Namespace() {
+  }
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-core/src/main/java/sonia/scm/repository/Namespace.java:62]: But then it would be public..?

**Pull request merged**